pkg/cloud/api/copy.go

```diff
@@ -0,0 +1,303 @@
+/*
+Copyright 2023 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package api
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+)
+
+type missingFieldOnCopy struct {
+	Path  Path
+	Value any
+}
+
+type copyContext struct {
+	// logSFn is an optional structured log function, matching the
+	// signature from klog/v2.
+	logSFn func(msg string, kv ...any)
+
+	missing []missingFieldOnCopy
+}
+
+func (c *copyContext) logS(msg string, kv ...any) {
+	if c.logSFn == nil {
+		return
+	}
+	c.logSFn(msg, kv...)
+}
+
+func copyValues(dest, src reflect.Value) (*copyContext, error) {
+	cc := &copyContext{}
+	err := copyValuesImpl(cc, Path{}, dest, src)
+	return cc, err
+}
+
+func copyValuesImpl(cc *copyContext, p Path, dest, src reflect.Value) error {
```

Comment:
nit: the `cc *copyContext` on all the methods tells me these should probably be struct member functions

Reply:
true -- was debating this, I can flip if you think this works better.